File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/managers/StatisticsManager.scala
package edu.uci.ics.amber.engine.architecture.worker.managers

import edu.uci.ics.amber.core.executor.{OperatorExecutor, SinkOperatorExecutor}
import edu.uci.ics.amber.engine.architecture.worker.statistics.{PortTupleMetrics, WorkerStatistics}
import edu.uci.ics.amber.core.workflow.PortIdentity

import scala.collection.mutable

class StatisticsManager {
  // DataProcessor
  private val inputTupleCount = mutable.Map.empty[PortIdentity, Long].withDefaultValue(0L)
  private val inputTupleSize = mutable.Map.empty[PortIdentity, Long].withDefaultValue(0L)
  private val outputTupleCount = mutable.Map.empty[PortIdentity, Long].withDefaultValue(0L)
  private val outputTupleSize = mutable.Map.empty[PortIdentity, Long].withDefaultValue(0L)
  private var dataProcessingTime: Long = 0L
  private var totalExecutionTime: Long = 0L
  private var workerStartTime: Long = 0L

  // AmberProcessor
  private var controlProcessingTime: Long = 0L

  def getStatistics(operator: OperatorExecutor): WorkerStatistics = {
    val (userFriendlyOutputTupleCount, userFriendlyOutputTupleSize) = operator match {
      case _: SinkOperatorExecutor => (inputTupleCount, inputTupleSize)
      case _                       => (outputTupleCount, outputTupleSize)
    }

    WorkerStatistics(
      inputTupleCount.map {
        case (portId, tupleCount) => PortTupleMetrics(portId, tupleCount)
      }.toSeq,
      inputTupleSize.map { case (portId, tupleSize) => PortTupleMetrics(portId, tupleSize) }.toSeq,
      userFriendlyOutputTupleCount.map {
        case (portId, tupleCount) => PortTupleMetrics(portId, tupleCount)
      }.toSeq,
      userFriendlyOutputTupleSize.map {
        case (portId, tupleSize) => PortTupleMetrics(portId, tupleSize)
      }.toSeq,
      dataProcessingTime,
      controlProcessingTime,
      totalExecutionTime - dataProcessingTime - controlProcessingTime
    )
  }

  def getInputTupleCount: Long = inputTupleCount.values.sum

  def getOutputTupleCount: Long = outputTupleCount.values.sum

  def increaseInputTupleCount(portId: PortIdentity): Unit = {
    inputTupleCount(portId) += 1
  }

  def increaseInputTupleSize(portId: PortIdentity, size: Long): Unit = {
    inputTupleSize(portId) += size
  }

  def increaseOutputTupleCount(portId: PortIdentity): Unit = {
    outputTupleCount(portId) += 1
  }

  def increaseOutputTupleSize(portId: PortIdentity, size: Long): Unit = {
    outputTupleSize(portId) += size
  }

  def increaseDataProcessingTime(time: Long): Unit = {
    dataProcessingTime += time
  }

  def increaseControlProcessingTime(time: Long): Unit = {
    controlProcessingTime += time
  }

  def updateTotalExecutionTime(time: Long): Unit = {
    totalExecutionTime = time - workerStartTime
  }

  def initializeWorkerStartTime(time: Long): Unit = {
    workerStartTime = time
  }
}

File: core/amber/src/main/python/core/architecture/managers/statistics_manager.py
import typing

from proto.edu.uci.ics.amber.engine.common import PortIdentity
from proto.edu.uci.ics.amber.engine.architecture.worker import WorkerStatistics


class StatisticsManager:
    def __init__(self):
        self._input_tuple_count = {}
        self._output_tuple_count = {}
        self._data_processing_time = 0
        self._control_processing_time = 0
        self._total_execution_time = 0
        self._worker_start_time = 0

    def get_statistics(self):
        return WorkerStatistics(
            self._input_tuple_count,
            self._output_tuple_count,
            self._data_processing_time,
            self._control_processing_time,
            self._total_execution_time
            - self._data_processing_time
            - self._control_processing_time,
        )

    def increase_input_tuple_count(self, port_id: PortIdentity) -> None:
        if port_id is None:
            port = 0
        else:
            port = port_id.id

        if port not in self._input_tuple_count:
            self._input_tuple_count[port] = 0
        self._input_tuple_count[port] += 1

    def increase_output_tuple_count(self, port_id: PortIdentity = None) -> None:
        # Currently, the number of output ports is fixed to 1
        if port_id is None:
            port = 0
        else:
            port = port_id.id

        if port not in self._output_tuple_count:
            self._output_tuple_count[port] = 0
        self._output_tuple_count[port] += 1

    def increase_data_processing_time(self, time) -> None:
        self._data_processing_time += time

    def increase_control_processing_time(self, time) -> None:
        self._control_processing_time += time

    def update_total_execution_time(self, time) -> None:
        self._total_execution_time = time
